packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -2281,6 +2283,16 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       // "export declare" is equivalent to just "export".
       const isDeclare = this.eatContextual("declare");
 
+      if (
+        isDeclare &&
+        (this.isContextual("declare") || !this.shouldParseExportDeclaration())
+      ) {
+        throw this.raise(
```

Comment:
Should we not throw here to make it recoverable?

Reply:
I'm not sure about how we could recover. What should the AST for `export declare foo` be?

Reply:
Yeah, you're right. Let's stick with throwing!